pkg/util/winutil/crashreport/crashreport.go

```diff
@@ -0,0 +1,144 @@
+		hive:    hive,
+		baseKey: key,
+	}
+	return wcr, nil
+}
+
+func (wcr *WinCrashReporter) generateReportedValue(wcs probe.WinCrashStatus) string {
+	return fmt.Sprintf("%s_%s", wcs.FileName, wcs.DateString)
+}
+
+func (wcr *WinCrashReporter) haveAlreadyReported(wcs probe.WinCrashStatus) bool {
+
+	newval := wcr.generateReportedValue(wcs)
+
+	k, err := registry.OpenKey(wcr.hive, wcr.baseKey, registry.QUERY_VALUE)
+	if err != nil {
+		// key not even there
+		return false
+	}
+	defer k.Close()
+	reportedval, _, err := k.GetStringValue(reportedKey)
+	if err != nil {
+		return false
+	}
+	if newval == reportedval {
+		return true
+	}
+	return false
+}
+
+func (wcr *WinCrashReporter) setReported(wcs probe.WinCrashStatus) {
+	newval := wcr.generateReportedValue(wcs)
+
+	k, _, err := registry.CreateKey(wcr.hive, wcr.baseKey, registry.ALL_ACCESS)
+	if err != nil {
+		// key not even there
+		return
+	}
+	defer k.Close()
+	// if we can't set the value, there's nothing we can do.  On next agent
+	// start, the same crash will be reported if the file is still there.
+	_ = k.SetStringValue(reportedKey, newval)
+}
+
+func (wcr *WinCrashReporter) handleStartupError(err error) error {
+	if retry.IsErrWillRetry(err) {
+		wcr.startupWarnCount++
+		// this is an expected error, the check run will occur before the system probe
+		// comes up.  However, it should resolve.  If the number of these exceeds a
+		// given threshold, report as an error anyway
+		if wcr.startupWarnCount > maxStartupWarnings {
+			return err
+		}
+		return nil
+	}
+	return err
+}
+
+// CheckForCrash uses the system probe crash module to check for a crash
+func (wcr *WinCrashReporter) CheckForCrash() (*probe.WinCrashStatus, error) {
+	if wcr.hasRunOnce {
+		return nil, nil
+	}
+	sysProbeUtil, err := process_net.GetRemoteSystemProbeUtil(
+		dd_config.SystemProbe.GetString("system_probe_config.sysprobe_socket"))
+	if err != nil {
+		return nil, wcr.handleStartupError(err)
+	}
+
+	data, err := sysProbeUtil.GetCheck(sysconfig.WindowsCrashDetectModule)
+	if err != nil {
+		return nil, wcr.handleStartupError(err)
+	}
+	crash, ok := data.(probe.WinCrashStatus)
+	if !ok {
+		return nil, fmt.Errorf("Raw data has incorrect type")
+	}
+	/*
+	 * originally did this with a sync.once.  The problem is the check is run prior to the
+	 * system probe being successfully started.  This is OK; we just need to detect the BSOD
+	 * on first available run.
+	 *
+	 * so keep running the check until we are able to talk to system probe, after which
+	 * we don't need to run any more
+	 */
+	wcr.hasRunOnce = true
+	if !crash.Success {
+		return nil, fmt.Errorf("Error getting crash data %s", crash.ErrString)
+	}
+
+	if len(crash.FileName) == 0 {
+		// no crash data present.  this is actually good news.  We don't need to do anything
+		// else
+		return nil, nil
+	}
+
+	if wcr.haveAlreadyReported(crash) {
+		log.Infof("Not reporting event on already reported crash")
+		return nil, nil
+	}
+	wcr.setReported(crash)
```

Comment:
the crash hasn't been reported yet. Can we delay `hasRunOnce` and `setReported` until the telemetry sender returns success?

Reply:
we could.
But this is shared code (will also be used by the public facing, general purpose crash reporting check).
So this code is specifically to check for the crash.  This module is meant to minimize copy/paste with the other check.
If we're concerned about teh telemetry sender failing, then that handling would have to go in the check itself, not the shared code here.